File: js/balanced.recursive.js
const mapping = {
  '(': ')',
  '[': ']',
  '{': '}',
}

/**
 * @param {string} slice
 * @returns {boolean}
 */
function balanced(slice) {
  return expect(undefined, slice.split(""))
}

/**
 * @param {string|undefined} end
 * @param {string[]} chars
 * @returns {boolean}
 */
function expect(end, chars) {
  while (true) {
    let c = chars.shift();
    let good;
    switch (c) {
      case "(":
      case "{":
      case "[": {
        good = expect(mapping[c], chars);
        break;
      }
      case undefined:
      case ")":
      case "}":
      case "]": {
        return end === c;
      }
      default: {
        good = true // any other char
      }
    }
    if (!good) return false
  }
}

assert(balanced("[]"), true);
assert(balanced("["), false);
assert(balanced("(())"), true);
assert(balanced("((()"), false);
assert(balanced(")(())"), false);
assert(balanced("))))"), false);
assert(balanced("(()))("), false);
assert(balanced("([])"), true);
assert(balanced("([[[]]])"), true);
assert(balanced("([[[00]])"), false);
assert(balanced("([[[{0}]]])"), true);

function assert(lhs, rhs) {
  if (lhs !== rhs) {
    throw new Error(`${lhs} !== ${rhs}`);
  }
}

File: src/balanced_recursive.rs
use std::str::Chars;

fn balanced_recursive(input: &str) -> bool {
    expect(None, &mut input.chars())
}

fn expect(end: Option<char>, input: &mut Chars) -> bool {
    loop {
        let c = input.next();
        let good = match c {
            Some('(') => expect(Some(')'), input),
            Some('[') => expect(Some(']'), input),
            Some('{') => expect(Some('}'), input),
            Some(')') | Some(']') | Some('}') | None => {
                return end == c;
            }
            _ => true, // any other char
        };
        if !good {
            return false;
        }
    }
}

#[test]
fn test_balanced_recursive() {
    assert_eq!(balanced_recursive("[]"), true);
    assert_eq!(balanced_recursive("["), false);
    assert_eq!(balanced_recursive("(())"), true);
    assert_eq!(balanced_recursive("((()"), false);
    assert_eq!(balanced_recursive(")(())"), false);
    assert_eq!(balanced_recursive("))))"), false);
    assert_eq!(balanced_recursive("(()))("), false);
    assert_eq!(balanced_recursive("([])"), true);
    assert_eq!(balanced_recursive("([[[]]])"), true);
    assert_eq!(balanced_recursive("([[[00]])"), false);
    assert_eq!(balanced_recursive("([[[{0}]]])"), true);
}
